src/dmd/escape.d

```diff
@@ -1092,10 +1099,12 @@ bool checkNewEscape(Scope* sc, Expression e, bool gag)
                     TypeFunction tf = cast(TypeFunction)fd.type;
                     if (tf.isref)
                     {
-                        if (!gag)
-                            error(e.loc, "storing reference to outer local variable `%s` into allocated memory causes it to escape",
-                                  v.toChars());
-                        result = true;
+                        const(char)* msg = "storing reference to outer local variable `%s` into allocated memory causes it to escape";
+                        if (!gag && emitError)
+                            error(e.loc, msg, v.toChars());
+                        else if (!gag)
+                            deprecation(e.loc, msg, v.toChars());
```

Comment:
Notice the arguments to `error` and `deprecation` are exactly the same. This suggests that `emitError` could be replaced with a delegate (or function pointer) that is set to `error` or `deprecation` and passed instead. This pattern repeats 4 times.

Reply:
The delegate passed would also need to handle setting the `result` variable, which would be more involved than just hoisting the parameter.
The repetition is also due to the code being itself repetitive. I originally started to refactor it, but the diff was just too much.